src/samplers.jl

```diff
@@ -0,0 +1,151 @@
+
+
+function sampleSMC!(pc::ParticleContainer,resampler::Function =resample_systematic ,resampler_threshold::AbstractFloat = 0.5)
+
+    while consume(pc) != Val{:done}
+        ess = effectiveSampleSize(pc)
+        if ess <= resampler_threshold * length(pc)
+            # compute weights
+            Ws = weights(pc)
+            # check that weights are not NaN
+            @assert !any(isnan, Ws)
+            # sample ancestor indices
+            n = length(pc)
+            nresamples = n
+            indx = resampler(Ws, nresamples)
+            resample!(pc, indx)
+        end
+    end
+end
+
+# The resampler threshold is only imprtant for the first step!
+function samplePG!(pc::ParticleContainer,resampler::Function = resample_systematic ,ref_particle::Union{Particle,Nothing}=nothing , resampler_threshold =0.5)
```

Comment:
Here (and above in `sampleSMC!`) one shouldn't restrict `resampler` to `Function`s IMO. It's common to implement callable structs (especially for resamplers that might be interesting since it would allow to define cache arrays), which could not be used here.

Reply:
Thanks for that comment!

Reply:
Could we not make this step in a second PR?

Reply:
Sure